libfreerdp/core/gateway/arm.c

```diff
@@ -0,0 +1,348 @@
+
+	if (!s)
+		return FALSE;
+
+	sz = Stream_Length(s);
+
+	if (sz <= INT_MAX)
+		status = freerdp_tls_write_all(tls, Stream_Buffer(s), (int)sz);
+
+	Stream_Free(s, TRUE);
+	if (status >= 0 && content_length > 0)
+	{
+		status = freerdp_tls_write_all(tls, (const BYTE*)data, content_length);
+	}
+	return (status >= 0);
+}
+
+static void arm_free(rdpArm* arm)
+{
+	if (!arm)
+		return;
+
+	freerdp_tls_free(arm->tls);
+	http_context_free(arm->http);
+
+	free(arm);
+}
+
+BOOL arm_resolve_endpoint(rdpContext* context, DWORD timeout)
+{
+#ifndef WITH_CJSON
+	WLog_ERR(TAG, "arm gateway support not compiled in");
+	return FALSE;
+#else
+	if (!context)
+		return FALSE;
+
+	if ((context->settings->LoadBalanceInfoLength == 0) ||
+	    (context->settings->RemoteApplicationProgram == NULL))
+		return FALSE;
+
+	rdpArm* arm = (rdpArm*)calloc(1, sizeof(rdpArm));
+	char* lbi = NULL;
+	cJSON* json = NULL;
+	char* message = NULL;
+
+	if (arm)
+	{
+		arm->context = context;
+		arm->settings = arm->context->settings;
+		arm->tls = freerdp_tls_new(arm->settings);
+		HttpResponse* response = NULL;
+		long StatusCode;
+
+		if (!arm->tls)
+			goto arm_error;
+
+		arm->http = http_context_new();
+
+		if (!arm->http)
+			goto arm_error;
+
+		if (!http_context_set_uri(arm->http, "/api/arm/v2/connections/") ||
+		    !http_context_set_accept(arm->http, "application/json") ||
+		    !http_context_set_cache_control(arm->http, "no-cache") ||
+		    !http_context_set_pragma(arm->http, "no-cache") ||
+		    !http_context_set_connection(arm->http, "Keep-Alive") ||
+		    !http_context_set_user_agent(arm->http, "FreeRDP/3.0") ||
+		    !http_context_set_x_ms_user_agent(arm->http, "FreeRDP/3.0") ||
+		    !http_context_set_host(arm->http, arm->settings->GatewayHostname))
+			goto arm_error;
+
+		if (!arm_tls_connect(arm, arm->tls, timeout))
+			goto arm_error;
+
+		json = cJSON_CreateObject();
+		if (!json)
+			goto arm_error;
+		cJSON_AddStringToObject(json, "application", arm->settings->RemoteApplicationProgram);
+
+		lbi = calloc(arm->settings->LoadBalanceInfoLength + 1, sizeof(char));
+		if (!lbi)
+			goto arm_error;
+
+		memcpy(lbi, arm->settings->LoadBalanceInfo, arm->settings->LoadBalanceInfoLength);
+
+		cJSON_AddStringToObject(json, "loadBalanceInfo", lbi);
+		free(lbi);
+		lbi = NULL;
+
+		cJSON_AddNullToObject(json, "LogonToken");
+		cJSON_AddNullToObject(json, "gatewayLoadBalancerToken");
+
+		message = cJSON_PrintUnformatted(json);
+		if (!message)
+			goto arm_error;
+
+		if (!arm_send_http_request(arm, arm->tls, "POST", "application/json", message,
+		                           strlen(message)))
+			goto arm_error;
+
+		free(message);
+		message = NULL;
+		cJSON_Delete(json);
+		json = NULL;
+
+		response = http_response_recv(arm->tls, TRUE);
+		if (!response)
+			goto arm_error;
+
+		StatusCode = http_response_get_status_code(response);
+		if (StatusCode == HTTP_STATUS_OK)
```

Comment:
might be a good idea to split this function into smaller ones, one per scope (easier to read/understand)